lib/graphql/batch/execution_strategy.rb

```diff
@@ -1,57 +1,62 @@
 module GraphQL::Batch
   class ExecutionStrategy < GraphQL::Query::SerialExecution
-    attr_reader :batched_queries
-
-    def initialize
-      @batched_queries = Hash.new{ |hash, key| hash[key] = [] }
-    end
-
     class OperationResolution < GraphQL::Query::SerialExecution::OperationResolution
       def result
-        result = super
-        until execution_strategy.batched_queries.empty?
-          queries = execution_strategy.batched_queries.shift.last
-          queries.first.class.execute(queries)
-        end
-        result
+        GraphQL::Batch.wait(super)
+      ensure
+        GraphQL::Batch::Executor.current.clear
       end
     end
 
     class SelectionResolution < GraphQL::Query::SerialExecution::SelectionResolution
       def result
-        result_hash = super
-        result_hash.each do |key, value|
-          if value.is_a?(FieldResolution)
-            value.result_hash = result_hash
+        lazy_resolve_object(super)
+      end
+
+      private
+
+      def lazy_resolve_object(obj)
+        queries = nil
+        case obj
+        when Array
+          array = obj
+          array.each_with_index do |value, i|
+            if query = lazy_key_assign(array, i, value)
+              (queries ||= []) << query
+            end
+          end
+        when Hash
+          hash = obj
+          hash.each do |key, value|
+            if query = lazy_key_assign(hash, key, value)
+              (queries ||= []) << query
+            end
+          end
+        end
+        queries ? GraphQL::Batch.all(queries).then { obj } : obj
+      end
+
+      def lazy_key_assign(obj, key, value)
+        value = lazy_resolve_object(value)
+        if value.is_a?(Promise)
+          obj[key] = nil
+          query = value.then do |result|
+            obj[key] = result
           end
         end
-        result_hash
+        query
       end
     end
 
     class FieldResolution < GraphQL::Query::SerialExecution::FieldResolution
-      attr_accessor :result_hash
-
       def get_finished_value(raw_value)
-        if raw_value.is_a?(QueryContainer)
-          raw_value.query_listener = self
-          register_queries(raw_value)
-          self
+        if raw_value.is_a?(Promise)
+          raw_value.then(->(result) { super(result) },
+                         ->(exc) { exc.is_a?(GraphQL::ExecutionError) ? super(exc) : raise(exc) })
```

Comment:
Oddly graphql-ruby expects a GraphQL::ExecutionError to be returned rather than raised (https://github.com/rmosolgo/graphql-ruby/blob/v0.10.9/lib/graphql/query/serial_execution/field_resolution.rb#L30).  @rmosolgo why doesn't graphql-ruby allow a `GraphQL::ExecutionError` to be raised from a resolve proc to specify an error?

There is support for doing exception handing in a middleware, but the middleware is used in `get_raw_value`, which has already returned, so isn't very useful for error handling in promises.